src/glb-director-xdp/main.go

```diff
@@ -342,6 +348,82 @@ func (app *Application) ReloadForwardingTable() {
+		sum.Encapsulated += curr[cpuIndex].Encapsulated - last[cpuIndex].Encapsulated
+
+		sum.UnknownFormat += curr[cpuIndex].UnknownFormat - last[cpuIndex].UnknownFormat
+		sum.NoMatchingBind += curr[cpuIndex].NoMatchingBind - last[cpuIndex].NoMatchingBind
+		sum.Matched += curr[cpuIndex].Matched - last[cpuIndex].Matched
+
+		sum.ErrorTable += curr[cpuIndex].ErrorTable - last[cpuIndex].ErrorTable
+		sum.ErrorSecret += curr[cpuIndex].ErrorSecret - last[cpuIndex].ErrorSecret
+		sum.ErrorHashConfig += curr[cpuIndex].ErrorHashConfig - last[cpuIndex].ErrorHashConfig
+		sum.ErrorMissingRow += curr[cpuIndex].ErrorMissingRow - last[cpuIndex].ErrorMissingRow
+		sum.ErrorCreatingSpace += curr[cpuIndex].ErrorCreatingSpace - last[cpuIndex].ErrorCreatingSpace
+		sum.ErrorMissingGatewayMAC += curr[cpuIndex].ErrorMissingGatewayMAC - last[cpuIndex].ErrorMissingGatewayMAC
+		sum.ErrorMissingSourceAddress += curr[cpuIndex].ErrorMissingSourceAddress - last[cpuIndex].ErrorMissingSourceAddress
+	}
+
+	return sum
+}
+
+func (app *Application) RunStatsCollection() {
+	globalCounters := app.Collection.Maps["glb_global_packet_counters"]
+	if globalCounters == nil {
+		log.Fatal("Could not load map glb_global_packet_counters")
+	}
+
+	var lastGlobalValues []C.struct_glb_global_stats
+
+	for {
+		var globalValues []C.struct_glb_global_stats
+
+		err := globalCounters.Lookup(uint32(0), &globalValues)
+		if err == nil {
+			if len(lastGlobalValues) > 0 {
+				sum := diffAndSumGlobalStats(lastGlobalValues, globalValues)
+				app.StatsClient.Count("packets.processed", int64(sum.Processed), nil, 1)
+				app.StatsClient.Count("packets.encapsulated", int64(sum.Encapsulated), nil, 1)
+
+				app.StatsClient.Count("packets.results", int64(sum.UnknownFormat), []string{"result:UnknownFormat"}, 1)
+				app.StatsClient.Count("packets.results", int64(sum.NoMatchingBind), []string{"result:NoMatchingBind"}, 1)
+				app.StatsClient.Count("packets.results", int64(sum.Matched), []string{"result:Matched"}, 1)
+
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorTable), []string{"error:ErrorTable"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorSecret), []string{"error:ErrorSecret"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorHashConfig), []string{"error:ErrorHashConfig"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorMissingRow), []string{"error:ErrorMissingRow"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorCreatingSpace), []string{"error:ErrorCreatingSpace"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorMissingGatewayMAC), []string{"error:ErrorMissingGatewayMAC"}, 1)
+				app.StatsClient.Count("packets.internal_errors", int64(sum.ErrorMissingSourceAddress), []string{"error:ErrorMissingSourceAddress"}, 1)
+			}
+
+			lastGlobalValues = globalValues
+		}
+
+		time.Sleep(10 * time.Second)
```

Comment:
Fairly minor, but in cases where the stats emission takes a bit of time it'll get added to the 10s here and things will skew a bit each cycle rather than being on 10s intervals. That can eventually cause funkiness in what the statsd workalike sees as things alias with its time periods. Probably best to do a `start` at the top of the `for` block and then `time.Sleep(10s - now - start)`.

Reply:
Looks like `time.Tick` is the standard way of doing this, in which the loop runs when the ticker sends on the channel, keeping it independent of the time the task itself takes. I'll switch to this, which seems to reuse that logic from the standard library.